docs/04-reference/winglang-spec.md

```diff
@@ -782,30 +782,153 @@ let f = (arg1: num, var arg2: num) => {
 
 ### 1.7 Optionality
 
-Symbol `?` can mark a type as optional.  
-Optionality means the value behind type can be either present or nil.
+Tri-state "nullity" values (`null`, `undefined` in JavaScript) are a primary source of bugs in
+software. Being able to guarantee that a value will never be null makes it possible to write safe
+code without thinking about nullity.
 
-Rules of optionality apply to the entire new container type of `type?` and not
-the value behind it (`type`).  
+So we are going to try and design Wing without allowing users to explicitly assign or use the `nil`
+value. Instead, the language offers syntax to support the various use cases that arise when dealing
+with *optionality*.
 
-Using the `??` operator allows one to safely access the value behind an optional
-variable by always providing a default, in case the variable is nil. This forces
-a value to be present for the l-value (left hand side of the assignment operator
-and guarantees what's returned is type-stripped from its `?` keyword).
+#### 1.7.1 Definition
 
-> ```TS
-> let x? = 44;
-> let y = x ?? 55;
-> ```
+The symbol `?` can mark a variable, function argument, return type or a field as optional (same as
+`Option<T>` in Rust).
 
-<details><summary>Equivalent TypeScript Code</summary>
+```js
+struct Person {
+  name: str;
+  address: str?;
+}
 
-> ```TS
-> const x: number? = 44;
-> const y: number = x ?? 55;
+let x = (foo: str?): num? => {
+  print(foo ?? "foo is not defined");
+  return 12;
+};
+
+class Foo {
+  my_opt: num?;
+}
+```
+
+In the above struct, the `address` field, the `foo` argument and the `my_opt` field are all marked
+as optionals using `?`.
+
+This means, for example, that when a `Person` can be initialized without defining the `address`
+field:
+
+```js
+let my_person = Person { name: "david" };
+```
+
+Let's look at various use cases for how to interact with optionals.
+
+#### 1.7.2 Assignment
+
+A reassignable variable or field can only be assigned to a concrete value. There is no way to
+explicitly assign a "null" value:
+
+```js
+let var x: str?;
+x = "hello"; // OK
+
+x = nil;
+//  ^^^ ERROR: no explicit `nil` in this language.
+```
+
+> You can be a smart-ass and do this:
+> ```js
+> let make_nil = (): str? => { return ?; }
+> x = make_nil();
```

Comment:
Why not support assigning a "not assigned" value:
```
let var x: num? = 7;
x = ?;
```
I guess we can play around with not supporting this and only allowing the smart-ass solution above, for a start. WDYT?

Reply:
yes, that was my thinking. let's try without this and see what blows up.